src/ReadReceiptsModel.cpp

```diff
@@ -0,0 +1,118 @@
+// SPDX-FileCopyrightText: 2021 Nheko Contributors
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "ReadReceiptsModel.h"
+
+#include <QLocale>
+
+#include "Cache.h"
+#include "Cache_p.h"
+#include "Logging.h"
+#include "Utils.h"
+
+ReadReceiptsModel::ReadReceiptsModel(QString event_id, QString room_id, QObject *parent)
+  : QAbstractListModel{parent}
+  , event_id_{event_id}
+  , room_id_{room_id}
+{
+        try {
+                addUsers(cache::readReceipts(event_id_, room_id_));
+        } catch (const lmdb::error &) {
+                nhlog::db()->warn("failed to retrieve read receipts for {} {}",
+                                  event_id_.toStdString(),
+                                  room_id_.toStdString());
+
+                return;
+        }
+
+        connect(cache::client(), &Cache::newReadReceipts, this, &ReadReceiptsModel::update);
+}
+
+void
+ReadReceiptsModel::update()
+{
+        try {
+                addUsers(cache::readReceipts(event_id_, room_id_));
+        } catch (const lmdb::error &) {
+                nhlog::db()->warn("failed to retrieve read receipts for {} {}",
+                                  event_id_.toStdString(),
+                                  room_id_.toStdString());
+
+                return;
+        }
+}
+
+QHash<int, QByteArray>
+ReadReceiptsModel::roleNames() const
+{
+        return {{Mxid, "mxid"},
+                {DisplayName, "displayName"},
+                {AvatarUrl, "avatarUrl"},
+                {Timestamp, "timestamp"}};
+}
+
+QVariant
+ReadReceiptsModel::data(const QModelIndex &index, int role) const
+{
+        if (!index.isValid() || index.row() >= (int)readReceipts_.size() || index.row() < 0)
+                return {};
+
+        switch (role) {
+        case Mxid:
+                return readReceipts_[index.row()].first;
+        case DisplayName:
+                return cache::displayName(room_id_, readReceipts_[index.row()].first);
+        case AvatarUrl:
+                return cache::avatarUrl(room_id_, readReceipts_[index.row()].first);
+        case Timestamp:
+                return dateFormat(readReceipts_[index.row()].second);
+        default:
+                return {};
+        }
+}
+
+void
+ReadReceiptsModel::addUsers(
+  const std::multimap<uint64_t, std::string, std::greater<uint64_t>> &users)
+{
+        auto oldLen = readReceipts_.length();
+
+        beginInsertRows(QModelIndex{}, oldLen, users.size() - 1);
```

Comment:
This won't work. You need to pass exactly where you are inserting elements. Not just assume they will land at the end of the list and then sort it.

Reply:
The existing count of elements is updated about 2 lines below that

Reply:
This line says you are inserting the elements at the end. If you call sort, that is not necessarily true.

Reply:
If you have any suggestions for how to do this better, I would be glad to hear them though

Reply:
> This won't work. You need to pass exactly where you are inserting elements. Not just assume they will land at the end of the list and then sort it.

No, what I have does work. It first inserts more receipts. Then it calls `dataChanged` on the number of receipts that were previously there. (i.e. there are receipts a and b. We add c and d. The model adds 2 items, although the order is a, c, b, d for some reason. Now Qt sees the model as a, b, b, d. But the model updates the first 2 (existing) items, so Qt sees a,c, b, d.)

Reply:
Well, then just call resetModel instead of insert + dataChanged. But really, you should probably call insert at the right places?

Reply:
If you don't want to write that logic yourself, put a sortproxy in front of the model.